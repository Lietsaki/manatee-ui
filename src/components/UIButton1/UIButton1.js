import React from 'react';
import { string, object, func } from 'prop-types';
import styled from 'styled-components';
import { getHSLA, getHSLADarker, getBeforeDimensions } from '@bit/lietsaki.manatee-ui.manatee-utilities';

const defaultColor = { hue: '211', saturation: '3', lightness: '60', alpha: '1' };
const defaultGradientColor1 = '#659999';
const defaultGradientColor2 = '#f4791f';

const UIButton1 = styled.button`
  color: ${(props) => (props.textColor ? props.textColor : '#fff')};
  background-color: ${(props) => (props.bgColor ? getHSLA(props.bgColor) : getHSLA(defaultColor))};
  width: ${(props) => (props.width ? props.width : '120px')};
  height: ${(props) => (props.height ? props.height : '37px')};
  border-radius: 10rem;
  display: flex;
  justify-content: center;
  align-items: center;
  transition: all 0.3s ease-in-out;
  border: none;
  font-size: 15px;
  outline: none;

  &:hover {
    cursor: pointer;
    background-color: ${(props) => (props.bgColor ? getHSLADarker(props.bgColor) : getHSLADarker(defaultColor))};
  }

  &:active {
    transition: all 0.1s ease-in-out;
    transform: scale(0.97);
  }

  &:before {
    content: '';
    display: block;
    position: absolute;
    z-index: -20;
    width: ${(props) =>
      props.width ? getBeforeDimensions('width', props.width) : getBeforeDimensions('width', '120px')};
    height: ${(props) =>
      props.height ? getBeforeDimensions('height', props.height) : getBeforeDimensions('height', '37px')};
    border-radius: 2rem;
    transition: box-shadow 0.2s;
    will-change: box-shadow;
  }

  &:active:before,
  &:focus:before {
    box-shadow: -0.09375rem -0.09375rem 0 0.09375rem ${(props) => (props.beforeGradient1 ? props.beforeGradient1 : defaultGradientColor1)},
      0 0 0 0.1875rem ${(props) => (props.beforeGradient2 ? props.beforeGradient2 : defaultGradientColor2)};
  }
`;

const GetUIButton1 = (props) => {
  const { onClick, bgColor, textColor, width, height, beforeGradient1, beforeGradient2 } = props;

  // Trigger event by pressing the enter key.
  const handleKeypress = (e) => {
    if (e.keyCode === 13) {
      onClick();
    }
  };

  return (
    <UIButton1
      onClick={onClick}
      onKeyPress={handleKeypress}
      bgColor={bgColor}
      textColor={textColor}
      width={width}
      height={height}
      beforeGradient1={beforeGradient1}
      beforeGradient2={beforeGradient2}
    >
      {props.children}
    </UIButton1>
  );
};

GetUIButton1.defaultProps = {
  bgColor: defaultColor,
  beforeGradient1: defaultGradientColor1,
  beforeGradient2: defaultGradientColor2,
  textColor: '#fff',
  width: '120px',
  height: '37px'
};

GetUIButton1.propTypes = {
  onClick: func,
  bgColor: object,
  textColor: string,
  width: string,
  height: string,
  beforeGradient1: string,
  beforeGradient2: string
};

export default GetUIButton1;
