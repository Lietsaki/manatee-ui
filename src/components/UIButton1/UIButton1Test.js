import React from 'react';
import UIButton1 from './UIButton1';

// To test all our customization options, let's create a button using every possible prop.
const dangerRed = { hue: '349', saturation: '79', lightness: '50', alpha: '1' };
const logClick = () => console.log('I have been clicked!');

const UIButton1Test = () => {
  return (
    <>
      <UIButton1>
        <p>Cancel</p>
      </UIButton1>
      <UIButton1
        onClick={logClick}
        bgColor={dangerRed}
        textColor="#151515"
        width="15vw"
        height="5vh"
        beforeGradient1="#FF416C"
        beforeGradient2="#FF4B2B"
      >
        <p>Sign In</p>
      </UIButton1>
    </>
  );
};

export default UIButton1Test;
