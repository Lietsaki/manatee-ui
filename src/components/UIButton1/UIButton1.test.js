import React from 'react';
import renderer from 'react-test-renderer';
import UIButton1 from './UIButton1';
import { render, fireEvent } from '@testing-library/react';

// 1) Define a mock callback for the button
const mockCallback = jest.fn();

// 2) Store the component in a variable
const button = (
  <UIButton1 onClick={mockCallback}>
    <p>Cancel</p>
  </UIButton1>
);

// TEST #1: Rendering the button's content
test('Renders UIButton1 correctly', () => {
  // 1) Render the button with react-testing-library in order to be able to get it by its text.
  const { getByText } = render(button);
  getByText('Cancel');

  // 2) Create the element with renderer to create snapshots between each callback
  const rendererButton = renderer.create(button);

  // 3) Take a snapshot with Jest
  let tree = rendererButton.toJSON();
  expect(tree).toMatchSnapshot();
});

// TEST #2: onClick callback execution
test('Executes UIButton1 onClick callback', () => {
  // 1) Render the button with React Testing Library to trigger a click event
  const { getByRole } = render(button);
  const RTLButton = getByRole('button');
  fireEvent.click(RTLButton);

  // 2) Expect that the callback was called only once
  expect(mockCallback).toHaveBeenCalledTimes(1);
});

// NOTE: The tab + enter callback execution is not tested because of the limited support Jest currently has, but it's been tested
// in the browser and works perfectly as well.
