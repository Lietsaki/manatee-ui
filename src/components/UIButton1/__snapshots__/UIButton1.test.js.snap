// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders UIButton1 correctly 1`] = `
<button
  className="sc-bdfBwQ hsQBxA"
  height="37px"
  onClick={[MockFunction]}
  onKeyPress={[Function]}
  width="120px"
>
  <p>
    Cancel
  </p>
</button>
`;
